### ISSUE TITLE
Retraining the diffusion, not able to reproduce the result

### ISSUE THREAD

**Issue by andyl-flwls** [RobinWitch/SynTalker#3]
Hi,

I find out I am not able to reproduce the training result. While testing the downloaded checkpoint, I obtain a FID of 0.465, which eactly matches the paper report. Can you share more about the training logs and previous checkpoint if possible.

Here is my evaluation for every 100 epoches:

300                                                                                                                                                                                                                                                            
fid score: 0.9149551874524606                                                                                                                       
align score: 0.7228679958479756                                                                                                                     
l1div score: 10.8033800663313

400
fid score: 0.6336881169913404
align score: 0.7410722973202462
l1div score: 11.488939134584863

500:
fid score: 0.7479078395751007
align score: 0.7385565538599137
l1div score: 12.119891646753315

600:
fid score: 0.5567867920079124
align score: 0.7389808562734564
l1div score: 12.484624376146304

700:
fid score: 0.5409238718580092
align score: 0.7433599146436639
l1div score: 12.868590413044053

800:
fid score: 0.589655917413082
align score: 0.7418395229904172
l1div score: 12.87576659054035

900:
fid score: 0.5837550245100491
align score: 0.7421088062676595
l1div score: 12.89307791061918

**Comment by RobinWitch** [comment]
Hello, andyl-flwls,

Thank you for trying to reproduce our work.
I have provided the training log and related training config in the ckpt download folder. If everything goes well, they should have been located under the folder you downloaded: `ckpt/beatx2_cospeech_diffusion`, with file names `0403_212319_diffusion_rvqvae_128.txt` and `0403_212319_diffusion_rvqvae_128.yaml`. More checkpoints can be downloaded from this [link](https://pan.quark.cn/s/9aaf7c9c12df).

Due to the dataset size and the feature extractor, fluctuations in FID are relatively normal. The inconsistency in retraining and evaluation results may be due to your current Python environment or hardware configuration. For example, with the same checkpoint, the difference in FID can reach up to 0.02 in different environments. Recently, we tested a result of 0.459 in the [Colab](https://colab.research.google.com/drive/1fEXgO-6wGYxfnFuqT12wUdk8-XpHx6yQ?usp=sharing) environment (which even exceeded the values in the paper).

From the evaluation log you provided, the result seems to be normal, and your FID also exceeds that of other methods. I think this is a great start. If you want to achieve a better result, you can try changing the training hyperparameters, such as the number of diffusion steps, learning rate, etc. Or try to add structures into the network to build a better relationships between speech and motion. I believe there is a lot of potential for improvement. Because in our work, we aimed to keep the network structure and training strategy as simple as possible (during the earlier experimental stages, we added some inductive bias, and the final FID reached up to 0.42).

**Comment by andyl-flwls** [comment]
Really appreciate your feedback. I checked the config and find out that for speaker id, you comment out the whole dataset but just left [2]. In the real experiment, should i change it back to all speakers or still rely on only [2]? The whole training dataset, after i change it back to all speakers, will be around 138GB, i just want to verify if my understanding is correct.

**Comment by RobinWitch** [comment]
Yes, to compare with other methods like emage, we follow the same setting, just using speaker2.

You are right, using all speakers will cost a lot space.

If you want to train all speakers, you may need to retrain rvqvae, because it is trained on a single speaker2.

Good luck!

**Comment by andyl-flwls** [comment]
thx for the information!

**Comment by andyl-flwls** [comment]
Dear authors,

I have in addition tried to use three A6000 card, one A100 and still find out unable to reproduce this result. The model performance will converge around 0.53 - 0.55 for various systems and computation cards. I keep all the training scripts and archectures to be the same but still am not able to reporduce this result. May I know if there could be some potential difference in the released code in the preprocessing or identity to be used as in the released code.

Thanks.

**Comment by RobinWitch** [comment]
Thank you again for trying out our work.

When obtaining the result, we used a single 4090 GPU, PyTorch version: 2.0.1+cu117, and CUDA version: 11.7. You can refer to the detailed configuration at [this link](https://huggingface.co/robinwitch/SynTalker/raw/main/ckpt/beatx2_cospeech_diffusion/0403_212319_diffusion_rvqvae_128.txt). The [checkpoints](https://pan.quark.cn/s/9aaf7c9c12df). generated during training have also been organized and provided to you earlier.

I'm certain that the preprocessing code is the same. If it were different, there could be serious issues with result reproducibility.
Regarding the identity aspect, following the BEAT2 benchmark, we only used the character 2.

We greatly appreciate you trying our work, and I'm not sure where your issue might lie. I think your current results are already quite good and outperform all other methods. I was able to stably reproduce the results when I was doing this project. Perhaps you should consider adjusting hyperparameters, such as the learning rate. Hyperparameters are crucial for all neural networks, especially for networks trained from scratch. There’s no need to strictly following our settings; I remember I trying a little few hyperparameter sets and get a FID improvement from 0.5x to 0.4x during my experiments. 

Best of luck!

**Comment by JiuFengSC** [comment]
> Hi,
> 
> I find out I am not able to reproduce the training result. While testing the downloaded checkpoint, I obtain a FID of 0.465, which eactly matches the paper report. Can you share more about the training logs and previous checkpoint if possible.
> 
> Here is my evaluation for every 100 epoches:
> 
> 300 fid score: 0.9149551874524606 align score: 0.7228679958479756 l1div score: 10.8033800663313
> 
> 400 fid score: 0.6336881169913404 align score: 0.7410722973202462 l1div score: 11.488939134584863
> 
> 500: fid score: 0.7479078395751007 align score: 0.7385565538599137 l1div score: 12.119891646753315
> 
> 600: fid score: 0.5567867920079124 align score: 0.7389808562734564 l1div score: 12.484624376146304
> 
> 700: fid score: 0.5409238718580092 align score: 0.7433599146436639 l1div score: 12.868590413044053
> 
> 800: fid score: 0.589655917413082 align score: 0.7418395229904172 l1div score: 12.87576659054035
> 
> 900: fid score: 0.5837550245100491 align score: 0.7421088062676595 l1div score: 12.89307791061918

Hi @andyl-flwls and @RobinWitch, 

Thanks for your discussion on this topic. I'm trying to train the model on all 25 speakers from the BEAT2 dataset, but I'm encountering issues with performance. (According to EMAGE, only 25 speakers are intact for the data. )

**What I Tried:**  I can successfully reproduce the performance for a single speaker (speaker [2]), which is close to the paper's results.

`FID:0.47, Align: 0.74, Diversity: 11.6`

**The Problem:** When I switch to training on all 25 speakers, my performance drops significantly:

`FID: 0.93, Align: 0.64, Diversity: 11.28`

**My Question:** I saw your reported scores, which look much better:

`fid score: 0.5837550245100491 align score: 0.7421088062676595 l1div score: 12.89307791061918` 

Were these scores achieved by training on all 25 speakers?

If so, could you please share what steps you took? Specifically:
- Did you only change the speaker configuration from `[2]` to `[1,2,3,4,.....,30]`?
- Or did you have to implement speaker ID embeddings?

I noticed the SynTalker codebase doesn't appear to have speaker embeddings implemented by default, which I saw in EMAGE and seems necessary for a multi-speaker model to differentiate between speakers.

Any clarification on this would be extremely helpful.

Thanks!
